operator-framework-junit5/src/main/java/io/javaoperatorsdk/operator/junit/OperatorExtension.java

```diff
@@ -0,0 +1,274 @@
+
+import static io.javaoperatorsdk.operator.api.config.ControllerConfigurationOverrider.override;
+
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.api.model.KubernetesResourceList;
+import io.fabric8.kubernetes.api.model.NamespaceBuilder;
+import io.fabric8.kubernetes.client.DefaultKubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.dsl.NonNamespaceOperation;
+import io.fabric8.kubernetes.client.dsl.Resource;
+import io.fabric8.kubernetes.client.utils.KubernetesResourceUtil;
+import io.fabric8.kubernetes.client.utils.Utils;
+import io.javaoperatorsdk.operator.Operator;
+import io.javaoperatorsdk.operator.api.ResourceController;
+import io.javaoperatorsdk.operator.api.config.BaseConfigurationService;
+import io.javaoperatorsdk.operator.api.config.ConfigurationService;
+import io.javaoperatorsdk.operator.api.config.Version;
+import io.javaoperatorsdk.operator.processing.ConfiguredController;
+import io.javaoperatorsdk.operator.processing.retry.Retry;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import org.awaitility.Awaitility;
+import org.junit.jupiter.api.extension.AfterAllCallback;
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class OperatorExtension
+    implements HasKubernetesClient,
+    BeforeAllCallback,
+    BeforeEachCallback,
+    AfterAllCallback,
+    AfterEachCallback {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(OperatorExtension.class);
+
+  private final KubernetesClient kubernetesClient;
+  private final ConfigurationService configurationService;
+  private final Operator operator;
+  private final List<ControllerSpec> controllers;
+  private final boolean preserveNamespaceOnError;
+  private final boolean waitForNamespaceDeletion;
+
+  private String namespace;
+
+  private OperatorExtension(
+      ConfigurationService configurationService,
+      List<ControllerSpec> controllers,
+      boolean preserveNamespaceOnError,
+      boolean waitForNamespaceDeletion) {
+
+    this.kubernetesClient = new DefaultKubernetesClient();
+    this.configurationService = configurationService;
+    this.controllers = controllers;
+    this.operator = new Operator(this.kubernetesClient, this.configurationService);
+    this.preserveNamespaceOnError = preserveNamespaceOnError;
+    this.waitForNamespaceDeletion = waitForNamespaceDeletion;
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  @Override
+  public void beforeAll(ExtensionContext context) throws Exception {
+    before(context);
```

Comment:
Why do we execute this before all? Should be this either before each or before all but not both?

Reply:
the invocation depend on how you instantiate the extension:
- if static --> beforeAll 
- if non static --> beforeEach

(at least that is my understanding)

Reply:
ahh, ok, thank you!